<script src="../lib/fx.js"></script>


# Promise

## 일급

<script>
  function add10(a, callback) {
    setTimeout(() => callback(a + 10), 1000);
  }

  add10(5, res => {
    add10(res, res => {
      add10(res, res => {
        log(res);
      });
    });
  });

  function add20(a) {
    return new Promise(
      resolve => setTimeout(() => resolve(a + 20), 1000));
  }

  add20(5)
    .then(add20)
    .then(add20)
    .then(add20)
    .then(add20)
    .then(log)


    // Callback 패턴은 함수를 반환하여, Promise는 해당 객체의 값을 반환한다.
    // Promise의 장점은 callback패턴의 가독성을 보완해준다.
    // 그보다 중요한 점은 Promise는 callback과 비동기로 일어난 일에 대해, 값으로 다룰 수 있는 것이다.
    // 
</script>

## 일급활용
<script>
  const delay100 = arg => new Promise(res => setTimeout(() => res(arg)), 100)
  // const go1 = (arg, fn) => fn(arg);
  const add5 = arg => arg + 5;
  // let result1 = go1(10, add5);
  // let result2 = go1(delay100(10),add5);
  // log(result1);
  // log(result2.then(log));
  go1(go1(10, add5),log);
  go1(go1(delay100(10),add5),log);
  
</script>


## Composition
<script type="module">
  // f . g
  // f(g(x))

  const g = a => a + 1;
  const f = a => a * a;

  // log(f(g(1)));
  // log(f(g()));

  Array.of(1).map(g).map(f)/*.forEach(r => log(r))*/;
  [].map(g).map(f)/*.forEach(r => log(r))*/;

  //Array는 map,filter 등의 함수로 값을 합성한다.

  Promise.resolve(2).then(g).then(f).then(r => log(r));
  new Promise(resolve =>
    setTimeout(() => resolve(2), 100)
  ).then(g).then(f).then(r => log(r));

  //Promise는 then으로 값을 합성한다.
</script>

## Kleisli Composition

<script type="module">
  console.clear();
  var users = [
    { id: 1, name: 'aa' },
    { id: 2, name: 'bb' },
    { id: 3, name: 'cc' },
  ];

  const getUserById = id =>
    find(u => u.id == id, users);

  const findName = ({ name }) => name;
  const getUser = getUserById;
  const findNameById = id => findName(getUser(id));

  const result = findNameById(2);
  log(result);

  // 외부에 의한 변화가 이루어 질 경우
  users.pop();
  users.pop();
  const result2 = findNameById(2);
  log(result); // 에러 발생!

</script>

<script type="module">
  console.clear();
  // 위와 같은 상황에서 Promise를 활용한 에러 처리 방법
  var users = [
    { id: 1, name: 'aa' },
    { id: 2, name: 'bb' },
    { id: 3, name: 'cc' },
  ];

  const edit_getUserById = id =>
    find(u => u.id == id, users) || Promise.reject('없어요!');

  const edit_findName = ({ name }) => name;
  const edit_getUser = edit_getUserById;

  const edit_findNameById = id => Promise.resolve(id)
  .then(edit_getUser)
  .then(edit_findName)
  .catch(a => a);
  edit_findNameById(6).then(log);
</script>